### PR TITLE
SymCrypt provider initial integration

### PR DESCRIPTION
**Pull request by mamckee**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

This PR adds the SymCrypt provider for OpenSSL 3. It will be used by default if present and enabled. 
 
###### Change Log  

<ins>openssl</ins> 
- Re-adds the FIPS mode flag check patch to OpenSSL and modifies the previously removed forced FIPS patch to load the SymCrypt provider by default. 
    - In non-FIPS modeThe SymCrypt provider will be loaded as the default if present, but may be disabled in favor of other providers.
    - In FIPS mode, the SymCrypt provider will always be implicitly loaded.

<ins>SymCrypt</ins>
- Updates SymCrypt to 103.4.1 which includes changes needed by the SymCrypt provider.

<ins>SymCrypt-OpenSSL</ins>
- Updates the SymCrypt-OpenSSL to include the new SymCrypt provider for OpenSSL 3. 
    - The SymCrypt provider replaces the SymCrypt engine for OpenSSL 1.1.1, since the engine model will be deprecated.

###### Does this affect the toolchain?  

**YES/NO**

###### Test Methodology

- Local build

### CODE REVIEW

File: SPECS/openssl/0032-Force-fips.patch [hunk 1]
```diff
@@ -0,0 +1,82 @@
+From 97edba1a35e1cc40b8b700aef75dd8904bcb2f33 Mon Sep 17 00:00:00 2001
+From: Maxwell Moyer-McKee <mamckee@microsoft.com>
+Date: Thu, 25 Jan 2024 16:48:47 -0800
+Subject: [PATCH] Load the symcrypt provider by default if fips flag or
```
Comment (dmcilvaney) [inline SPECS/openssl/0032-Force-fips.patch:4]: What is the flow if the symcrypt packge is not installed, but fips is on? Out of scope for this PR but we have a validation tool for image configurations, and some documentation for FIPS, that might need to be updated if so.

File: SPECS/openssl/openssl.spec [hunk 2]
```diff
@@ -9,7 +9,7 @@
 Summary: Utilities from the general purpose cryptography library with TLS implementation
 Name: openssl
 Version: 3.1.4
-Release: 1%{?dist}
+Release: 2%{?dist}
```
Comment (tobiasb-ms) [inline SPECS/openssl/openssl.spec:12]: Let's leave it at Release 1 for now -- we'll lock the release in later on, but this way we can work incrementally without resulting in a huge release number at ship date.

**Review by tobiasb-ms — APPROVED** [review]

File: SPECS/openssl/0032-Force-fips.patch (hunk 1 above)
Comment (dmcilvaney) [reply SPECS/openssl/0032-Force-fips.patch:4]: I suppose it won't be without precedent... the system won't even boot without `dracut-fips` installed if you turn on fips mode....

Can we add this package to `imageconfigvalidator.go:validatePackages()` so we at least yell about it missing if fips is configured in the image json?

@tobiasb-ms We will want to update compliance and documentation to reflect this, or do we want to add a Requires on symcrpt?

Comment (dmcilvaney) [reply SPECS/openssl/0032-Force-fips.patch:4]: Had a chat offline, we can just ignore this for now and fix it later in the 3.0 release. Figure out what we want to do.

**Review by dmcilvaney — APPROVED** [review]

File: SPECS/SymCrypt-OpenSSL/SymCrypt-OpenSSL.spec [hunk 3]
```diff
@@ -34,26 +34,41 @@ mkdir bin; cd bin
...
-%{_includedir}/scossl.h
+%{_libdir}/engines-3/symcryptengine.so
+%{_libdir}/ossl-modules/symcryptprovider.so
+%{_includedir}/e_scossl.h
+%{_sysconfdir}/pki/tls/symcrypt_prov.cnf
 
 %changelog
+* Thu Dec 28 2023 Maxwell Moyer-McKee <mamckee@microsoft.com> - 1.4.0-1
+- Update SymCrypt-OpenSSL to v1.4.0.
+- Adds SymCryptp-OpenSSL provider for OpenSSL 3.
```
Comment (samuel-lee-msft) [inline SPECS/SymCrypt-OpenSSL/SymCrypt-OpenSSL.spec:70]: NP typo: SymCryptp-OpenSSL -> SymCrypt-OpenSSL

File: SPECS/SymCrypt/SymCrypt.spec [hunk 4]
```diff
@@ -40,15 +40,20 @@ A core cryptographic library written by Microsoft
...
 # Create a symbolic link as if jitterentropy-library has been pulled in as git submodule
-rm -rf jitterentropy-library
-ln -s jitterentropy-library-3.3.1 jitterentropy-library
+rm -rf 3rdparty/jitterentropy-library
+ln -s ../jitterentropy-library-3.3.1 3rdparty/jitterentropy-library
 
 %build
+SYMCRYPT_BRANCH=main \
+SYMCRYPT_COMMIT_HASH=c55c670 \
+SYMCRYPT_COMMIT_TIMESTAMP=2023-12-01T15:59:37-08:00 \
```
Comment (samuel-lee-msft) [inline SPECS/SymCrypt/SymCrypt.spec:49]: I think the hash and timestamp should reflect the hash and timestamp of the last commit in SymCrypt 103.4.1 just for clarity when these strings are built into the library:
SYMCRYPT_COMMIT_HASH=a84ffe1
SYMCRYPT_COMMIT_TIMESTAMP=2024-01-26T22:00:47-08:00

Comment (mamckee) [reply SPECS/SymCrypt/SymCrypt.spec:49]: Thanks for the catch. I forgot to update this after the release was published.

**Pull request merged**